V0.27  typo tolerance configuration (#1606)
meilisearch/documentation: learn/configuration/typo_tolerance.md

# Typo tolerance

Typo tolerance allows users to find documents they're looking for even when they make mistakes while typing. Meilisearch allows you to [configure the typo tolerance feature for each index](/reference/api/typo_tolerance.md#update-typo-tolerance).

:::note
Meilisearch considers a typo on a query's first character as two typos. This increases performance during search.

> :::

Reviewer: I think that this deserves more than a note, it is a first-class rule of typo tolerance

Author: We use note/warning containers to draw the readers' attention to important information so they don't scroll past it

:::note
Concatenating two strings will count as one typo.
:::

## Configuring typo tolerance

Typo tolerance is enabled by default, but you can disable it if needed:

<CodeSamples id="typo_tolerance_guide_1" />

Disabling typo tolerance will mean Meilisearch no longer applies typo tolerance to match queries at search time. However, it will still [sort results by increasing number of typos](#impact-of-typo-tolerance-on-the-typo-ranking-rule).

### `minWordSizeForTypos`

By default, Meilisearch applies typo tolerance to a query if its length is at least 5 characters. To accept two typos, the query should be at least 9 characters.

If your dataset contains `seven`, searching for `sevem` or `sevan` will match `seven`. But `tow` won't match `two` as it's less than 5 characters.

You can override these default settings using the `minWordSizeForTypos` object. The code sample below sets the minimum word size for one typo to 4 and the minimum word size for two typos to 10.

<CodeSamples id="typo_tolerance_guide_4" />

When updating the `minWordSizeForTypos` object, keep in mind that:

- the value of `twoTypos` should be greater or equal to `oneTypo`

> - the value for both `oneTypo` and `twoTypos` should be between `0` and `255`

Reviewer 2: For oneTypo:
If you put <= 2: f you have a very low value for oneTypo, you will get a lot of false positive results.
If you put >= 8: If you have a very high value for oneTypo, the typo tolerance feature of the search engine will be rarely used.

For twoTypos:
If you put <= 4: If you have a very small value for twoTypos, you will get many false positive results.
If you put >= 14: If you have a very high value for twoTypos, the search engine's typo tolerance feature will rarely match two typos.

Something like this seems reasonable, it is extracted from the error messages encountered when using Algolia. I think we can recommend something similar.

### `disableOnWords`

You can disable typo tolerance for a list of query terms by adding them to `disableOnWords`.

<CodeSamples id="typo_tolerance_guide_3" />

Meilisearch won't apply typo tolerance on the query term `Shrek` or `shrek` at search time to match documents.

:::note
`disableOnWords` is case insensitive.
:::

### `disableOnAttributes`

> You can disable the typo tolerance feature for a specific attribute by adding it to `disableOnAttributes`. The code sample below disables typo tolerance for `title`:

Reviewer: ```suggestion
You can disable the typo tolerance feature for a specific document attribute by adding it to `disableOnAttributes`. The code sample below disables typo tolerance for `title`:
```

<CodeSamples id="typo_tolerance_guide_2" />

> If you make any typos in `title` at search, Meilisearch won't match any documents.

Reviewer: It is a bit convoluted, feel free to simplify somehow, but this is actually how it works.
```suggestion
If you make a typo on a search term ("hallo" instead of "hello"), and "hello" is present in the `title` field of a document, then this occurrence of "hello" will not be a match at search time.
```

## Impact of typo tolerance on the `typo` ranking rule

The [`typo` ranking rule](/learn/core_concepts/relevancy.md#_2-typo)  sorts the results by increasing number of typos on matched query words. Documents with 0 typos will rank highest. This rule does not impact the typo tolerance setting.

If you don't use the `typo` ranking rule, but enable typo tolerance for an index, Meilisearch will use typo tolerance to match documents but won't sort them based on increasing number of typos.

## How are typos calculated

> Meilisearch uses the Levenshtein algorithm to check if words match. You can find more details in our [dedicated guide](/learn/advanced/levenshtein_algorithm.md).

Author: Does this section need more detail?

# Typo tolerance

Typo tolerance allows users to make mistakes when typing and still find results close to their query, e.g., typing `phnoe` instead of `phone`. Meilisearch allows you to [configure the typo tolerance feature for each index](/reference/api/typo_tolerance.md#update-typo-tolerance).

::: note
Meilisearch considers a typo on a query's first character as two typos. This increases performance during search.
:::

## Configuring typo tolerance

Typo tolerance is enabled by default, but you can disable it if needed:

<CodeSamples id="typo_tolerance_guide_1" />

Disabling typo tolerance will mean Meilisearch no longer applies typo tolerance to match queries at search time. However, it will still [sort results by increasing number of typos](#impact-of-typo-tolerance-on-the-typo-ranking-rule).

### `minWordSizeForTypos`

By default, Meilisearch applies typo tolerance to a query term if its length is at least `5` characters. The query term should be at least `9` characters long to accept two typos.

::: note
Concatenating two query terms will count as one typo.

> If you type `La tableau`, Meilisearch will concatenate it to `Letableau`. This concatenation will count as one typo. This rule applies to all [space separators](/learn/advanced/datatypes.md#string).

Reviewer 3: I think it's much more useful than you're imagining based off the example. Say someone searches for "hall way" or "any more"; they would probably also want to see results containing "hallway" or "anymore", and the same can be applied to numerous compound words in English.

> If you type `La tableau`, Meilisearch will concatenate it to `Letableau`. This concatenation will count as one typo. This rule applies to all [space separators](/learn/advanced/datatypes.md#string).

Reviewer 3: The example isn't great though, will try to improve it.

<CodeSamples id="typo_tolerance_guide_4" />

When updating the `minWordSizeForTypos` object, keep in mind that:

- the value of `twoTypos` should be greater or equal to `oneTypo`
- the value for both `oneTypo` and `twoTypos` should be between `0` and `255`

We recommend keeping the value of `oneTypo` between `2` and `8` and the value of `twoTypos` between `4` and `14`. If the value is too small, you will get a lot of false-positive results. A bigger value will mean the typo tolerance feature is rarely used to match documents.

### `disableOnWords`

You can disable typo tolerance for a list of query terms by adding them to `disableOnWords`.

<CodeSamples id="typo_tolerance_guide_3" />

Meilisearch won't apply typo tolerance on the query term `Shrek` or `shrek` at search time to match documents.

::: note
`disableOnWords` is case insensitive.
:::

### `disableOnAttributes`

You can disable the typo tolerance feature for a specific document attribute by adding it to `disableOnAttributes`. The code sample below disables typo tolerance for `title`:

<CodeSamples id="typo_tolerance_guide_2" />

> If you type `Dunbo` instead of `Dumbo` with the above settings, Meilisearch won't match any documents.

Reviewer 3: I used "Biutiful", the Innaritu movie. For future reference, "Pursuit of Happyness" is another good one, or "Inglourious Basterds" (less appropriate).

- insertion in _M_ of a character of _P_ (e.g., siting → sit**t**ing)
- deletion of a character from _M_ (e.g., satu**r**day → satuday)

By default, Meilisearch uses the following rules for matching documents, you can configure them using the [update typo tolerance endpoint](/reference/api/typo_tolerance.md#update-typo-tolerance). These rules are **by word** and not for the whole query string.

- If the query word is between `1` and `4` characters, **no typo** is allowed. Only documents that contain words that **start with** or are of the **same length** with this query word are considered valid
- If the query word is between `5` and `8` characters, **one typo** is allowed. Documents that contain words that match with **one typo** are retained for the next steps.
- If the query word contains more than `8` characters, we accept a maximum of **two typos**

This means that "saturday" which is `7` characters long, uses the second rule and matches every document containing **one typo**. For example:

- "saturday" is accepted because it is the same word
- "satuday" is accepted because it contains **one typo**

> - "s**u**tuday" is not accepted because it contains **two typos**

Reviewer 3: I find the bolding distracting in this case so I chose to remove it instead. (Also, consider that the second typo is not `u`, it's the missing `r`)

> - "**c**aturday" is not accepted because it contains **two typos**

Reviewer 4: I'm not opposed to having `caturday` as an example, but we it doesn't make sense to have both `satuday` and `caturday`.

Also, the description is wrong (caturday contains only one typo).

```suggestion
```

Reviewer 3: Disagree, `caturday` is an effective example because a typo on the first letter of a word counts as two typos.

<CodeSamples id="typo_tolerance_guide_1" />

Disabling typo tolerance will mean Meilisearch no longer applies typo tolerance to match queries at search time. However, it will still [sort results by increasing number of typos](#impact-of-typo-tolerance-on-the-typo-ranking-rule).

### `minWordSizeForTypos`

> By default, Meilisearch applies typo tolerance to a query if its length is at least 5 characters. To accept two typos, the query should be at least 9 characters.

Reviewer 2: ```suggestion
By default, Meilisearch applies typo tolerance to a query term if its length is at least 5 characters. To accept two typos, a query term should be at least 9 characters.
```

By reading this sentence one could understand that the type tolerance feature will be triggered according to the total size of the `q` field which is not true. It is the individual words in `q` on which the typo tolerance feature will be triggered.

`disableOnWords` is case insensitive.
:::

### `disableOnAttributes`

You can disable the typo tolerance feature for a specific attribute by adding it to `disableOnAttributes`. The code sample below disables typo tolerance for `title`:

<CodeSamples id="typo_tolerance_guide_2" />

If you make any typos in `title` at search, Meilisearch won't match any documents.

## Impact of typo tolerance on the `typo` ranking rule

> The [`typo` ranking rule](/learn/core_concepts/relevancy.md#_2-typo)  sorts the results by increasing number of typos on matched query words. Documents with 0 typos will rank highest. This rule does not impact the typo tolerance setting.

Reviewer 2: ```suggestion
The [`typo` ranking rule](/learn/core_concepts/relevancy.md#_2-typo) sorts the results by increasing number of typos on matched query words. Documents with 0 typos will rank highest. This rule does not impact the typo tolerance setting.
```
